### PR TITLE
Feature/mark custom types

### LINKED ISSUE #106: Mark methods on arbitrary classes
pydantic specific methods like `__get_validators__` should be marked with a TODO on arbitrary classes to make migration of custom types easier

### PR DESCRIPTION
**Pull request by JensHeinrich**
EDIT by @Kludex:
- Closes #106

### CODE REVIEW

**Comment by JensHeinrich** [comment]
- [ ] Rebase and reorder to make the "Fixes" commit the last commit after fixing the feature

**Comment by slafs** [comment]
Nice! Thank you. This will be useful for us too! 🙇 (haven't reviewed this PR though)

Just noticed that the rule code should probably be `BP009` (or `BP011` maybe). The last current working one is `BP008`, but `BP009` and `BP010` are only mentioned in the README, they are not-implemented and commented out in the README. Looks like the last time a new mod was added (BP008), Marcelo incremented the number of commented out mods and used first "free" one.

**Comment by JensHeinrich** [comment]
> Nice! Thank you. This will be useful for us too! 🙇 (haven't reviewed this PR though)
> 
> Just noticed that the rule code should probably be `BP009` (or `BP011` maybe). The last current working one is `BP008`, but `BP009` and `BP010` are only mentioned in the README, they are not-implemented and commented out in the README. Looks like the last time a new mod was added (BP008), Marcelo incremented the number of commented out mods and used first "free" one. See #90.

Ok, thought I would just append to the commented one^^

**Comment by JensHeinrich** [comment]
> Nice! Thank you. This will be useful for us too! 🙇 (haven't reviewed this PR though)
> 
> Just noticed that the rule code should probably be `BP009` (or `BP011` maybe). The last current working one is `BP008`, but `BP009` and `BP010` are only mentioned in the README, they are not-implemented and commented out in the README. Looks like the last time a new mod was added (BP008), Marcelo incremented the number of commented out mods and used first "free" one. See #90.

Updated the numbering

File: README.md [hunk 1]
```diff
@@ -301,7 +303,50 @@ class User(BaseModel):
     name: Annotated[str, StringConstraints(min_length=1)]
 ```
 
-<!-- ### BP009: Replace `pydantic.parse_obj_as` by `pydantic.TypeAdapter`
+### BP009: Mark pydantic "protocol" functions in custom types with proper TODOs
```
Comment (Kludex) [inline README.md:306]:
```suggestion
### BP009: Mark Pydantic "protocol" functions in custom types with proper TODOs
```

File: README.md [hunk 2]
```diff
@@ -301,7 +303,50 @@ class User(BaseModel):
...
+- ✅ Mark `__modify_schema__` as to be replaced by `__get_pydantic_json_schema__`.
+
+The following code will be transformed:
+
+```py
+class SomeThing:
+    @classmethod
+    def __get_validators__(cls):
+        yield from []
+        return
```
Comment (Kludex) [inline README.md:318]:
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -301,7 +303,50 @@ class User(BaseModel):
...
+
+    @classmethod
+    def __modify_schema__(
+        cls, field_schema: Dict[str, Any], field: Optional[ModelField]
+    ):
```
Comment (Kludex) [inline README.md:323]:
```suggestion
    def __modify_schema__(cls, field_schema, field):
```

File: README.md [hunk 4]
```diff
@@ -301,7 +303,50 @@ class User(BaseModel):
...
+    def __get_validators__(cls):
+        yield from []
+        return
+
+    @classmethod
+    # TODO[pydantic]: We couldn't refactor `__modify_schema__`, please create the `__get_pydantic_json_schema__` manually.
+    # Check https://docs.pydantic.dev/latest/migration/#defining-custom-types for more information.
+    def __modify_schema__(
+        cls, field_schema: Dict[str, Any], field: Optional[ModelField]
+    ):
```
Comment (Kludex) [inline README.md:344]:
```suggestion
    def __modify_schema__(cls, field_schema, field):
```

File: README.md [hunk 5]
```diff
@@ -368,6 +413,8 @@ class User(BaseModel):
     name: ImportString
 ``` -->
 
+
+
```
Comment (Kludex) [inline README.md:417]:
```suggestion
```

File: bump_pydantic/codemods/__init__.py [hunk 6]
```diff
@@ -31,6 +32,8 @@ class Rule(str, Enum):
     """Replace `@validator` with `@field_validator`."""
     BP008 = "BP008"
     """Replace `con*` functions by `Annotated` versions."""
+    BP009 = "BP009"
+    """Mark pydantic 'protocol' functions in custom types with proper TODOs"""
```
Comment (Kludex) [inline bump_pydantic/codemods/__init__.py:36]:
```suggestion
    '''Mark Pydantic "protocol" functions in custom types with proper TODOs.'''
```

**Review by Kludex — CHANGES_REQUESTED** [review]
We need to add a test for when there's already a TODO comment.

File: README.md [hunk 7]
```diff
@@ -301,7 +303,50 @@ class User(BaseModel):
...
+Into:
+
+```py
+class SomeThing:
+    @classmethod
+    # TODO[pydantic]: We couldn't refactor `__get_validators__`, please create the `__get_pydantic_core_schema__` manually.
+    # Check https://docs.pydantic.dev/latest/migration/#defining-custom-types for more information.
+    def __get_validators__(cls):
+        yield from []
+        return
```
Comment (Kludex) [inline README.md:337]:
```suggestion
```

**Review by Kludex — APPROVED** [review]

**Pull request merged**